Introducing support for azurerm_network_security_rule creation with for_each iteration meta-argument

juanjojulian opened the pull request:
## Describe your changes

Added support to create `azurerm_network_security_rule` resources with `for_each` instead of `count`

## Issue number

#90

## Checklist before requesting a review
- [X] The pr title can be used to describe what this pr did in `CHANGELOG.md` file
- [X] I have executed pre-commit on my machine
- [X] I have passed pr-check on my machine

Thanks for your cooperation!

lonegunmanb reviewed (changes requested):
Thanks @juanjojulian for opening this pr! Almost LGTM but I think maybe we can simplify the new added variable by changing the type to `bool`? WDYT?

juanjojulian commented on variables.tf:
```diff
@@ -25,6 +25,12 @@ variable "destination_address_prefixes" {
 
+variable "iteration" {
+  description = "Iteration technic to generate the rules, defaults to 'count' for compatibilty reasons, prefered method is 'for_each'"
+  type        = string
```
Totally agree, cleaner.

juanjojulian commented on main.tf:
```diff
@@ -59,4 +81,26 @@ resource "azurerm_network_security_rule" "custom_rules" {
+}
+
+resource "azurerm_network_security_rule" "custom_rules_for" {
+  for_each = { for value in var.custom_rules : value.name => value if var.iteration == "for_each" }
```
You are right, but this was also true for `count` as the `name` attribute has to be unique:

https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/network_security_rule
![CleanShot 2023-02-13 at 12 00 55](https://user-images.githubusercontent.com/43988825/218440761-c6d285d2-b53f-451b-a326-604b0143501d.png)

With the current configuration if you deploy more than one NSG rules without providing the name the module fails as all of them will be named 'default_rule_name':

`name = lookup(var.custom_rules[count.index], "name", "default_rule_name")`

I made 'name' required for both `count` and `for_each` and I took the opportunity to define a proper `type` for var.custom_rules. I mimic in the type definition what the 'lookups' were doing in the code.

Let me know what you think.

lonegunmanb replied:
Thanks for justification, I would consider removing such default value as a breaking change. Let's say we keep this behavior for now, and we open a new issue to record a TODO item to fix this issue in our next major version release, does that sounds ok to you? Whilst I'd like to warn our users that the rule's name must be a static hard coded string in the code, it cannot contains any part from another resource, like `random_string`.

lonegunmanb commented on main.tf:
```diff
@@ -35,21 +35,43 @@ resource "azurerm_network_security_rule" "predefined_rules" {
+  direction                                  = element(var.rules[lookup(each.value, "name")], 0)
+  name                                       = each.value.name
+  network_security_group_name                = azurerm_network_security_group.nsg.name
+  priority                                   = each.value.priority
```
I understand the reason why the `priority` in for_each version is different than it's counterpart in count version, because we're lacking of the index so we cannot calculate the priority when it's absent. We might need to warn our users the `priority` is required if you choose `var.for_each = true`.

lonegunmanb reviewed (changes requested):
Thanks @juaduan for the update, I have some concerns and I would prefer a small step rather then a pr which contains potential breaking changes. I think we can open some new issues to record these proposals that require breaking change and leave them to our next major version release. WDYT?

Btw, please rename `examples/For_each` to `examples/for_each`, thanks!

juanjojulian commented on variables.tf:
```diff
@@ -5,12 +5,27 @@ variable "resource_group_name" {
 
 # Custom security rules
-# [priority, direction, access, protocol, source_port_range, destination_port_range, description]"
+# [{name, priority, destination_port_range}]"
```
The comment also says `# All the fields are required.` but analysing the code you realise that only `name, priority, destination_port_range` are really required because the other attributes have defaults assigned. For instance:

`protocol = lookup(var.custom_rules[count.index], "protocol", "*")`
`access = lookup(var.custom_rules[count.index], "access", "Allow")`

juanjojulian commented:
for_each example was not deploying any kind of rule and failed the e2e test, added one rule to `custom_rules`.

lonegunmanb reviewed (changes requested):
Thanks @juanjojulian for the update, would you please rename `examples/For_each` to `examples/for_each`?

And thanks for pointing out, we also need to update our e2e tests, e.g.:

```go
func TestExamples(t *testing.T) {
	directories, err := os.ReadDir("../../examples")
	if err != nil {
		t.Fatal(err)
	}

	for _, d := range directories {
		directory := d
		if strings.HasPrefix(directory.Name(), "_") || !directory.IsDir() {
			continue
		}
		useForEach := []bool{
			false,
			true,
		}
		for _, f := range useForEach {
			t.Run(fmt.Sprintf("%s-%t", directory.Name(), f), func(t *testing.T) {
				test_helper.RunE2ETest(t, "../../", fmt.Sprintf("examples/%s", directory.Name()), terraform.Options{
					Upgrade: true,
					Vars: map[string]interface{}{
						"use_for_each": f,
					},
				}, func(t *testing.T, output test_helper.TerraformOutput) {
					nsgId, ok := output["network_security_group_id"].(string)
					assert.True(t, ok)
					assert.Regexp(t, "/subscriptions/.+/resourceGroups/.+/providers/Microsoft.Network/networkSecurityGroups/.+", nsgId)
				})
			})
		}
	}
}
```

We need update both e2e test and upgrade test like this way, and add a new variable `use_for_each` for all examples.

juanjojulian commented:
> Thanks @juanjojulian for the update, would you please rename `examples/For_each` to `examples/for_each`?
> 

I forgot that MacOS X default installation formats the disk with a case insensitive file system... No problem, fixed and committed.

lonegunmanb commented on test/e2e/terraform_network_security_group_test.go:
```diff
@@ -22,14 +22,23 @@ func TestExamples(t *testing.T) {
+			true,
+		}
+		for _, f := range useForEach {
+			t.Run(directory.Name(), func(t *testing.T) {
```
As we're running all our tests in parallel, would you please add the `userForEach`'s value to sub-test's name like this way?:

```go
t.Run(fmt.Sprintf("%s-%t", directory.Name(), f), func(t *testing.T) {
```

lonegunmanb commented on test/upgrade/upgrade_test.go:
```diff
@@ -28,10 +28,19 @@ func TestUpgrade(t *testing.T) {
+			true,
+		}
+		for _, f := range useForEach {
+			t.Run(directory.Name(), func(t *testing.T) {
```
As we're running all our tests in parallel, would you please add the `userForEach`'s value to sub-test's name like this way?:

```go
t.Run(fmt.Sprintf("%s-%t", directory.Name(), f), func(t *testing.T) {
```

lonegunmanb reviewed (changes requested):
Thanks @juanjojulian for the update, we might need to update sub tests' names, and we've missed `modules/CouchDB/main.tf`.

lonegunmanb reviewed (changes requested):
Thanks @juanjojulian for the update, we've got failed tests, one exmaple is:

```text
TestExamples/MySQL-true 2023-02-17T01:56:26Z retry.go:99: Returning due to fatal error: FatalError{Underlying: error while running command: exit status 1; ╷
│ Error: Unsupported attribute
│ 
│   on ../../main.tf line 45, in resource "azurerm_network_security_rule" "predefined_rules_for":
│   45:   priority                                   = each.value.priority
│     ├────────────────
│     │ each.value is object with 1 attribute "name"
│ 
│ This object does not have an attribute named "priority".
╵}
```

I guess we need update all example's default variables, and we might need an extra `precondition` to ensure that when `var.use_for_each` is `true`, then the `priority` must be set, this `precondition` could bring the users a much clearer message on what's wrong with their variables. WDYT?

juanjojulian commented:
> Thanks @juanjojulian for the update, we've got failed tests, one exmaple is:
> 
> ```
> TestExamples/MySQL-true 2023-02-17T01:56:26Z retry.go:99: Returning due to fatal error: FatalError{Underlying: error while running command: exit status 1; ╷
> │ Error: Unsupported attribute
> │ 
> │   on ../../main.tf line 45, in resource "azurerm_network_security_rule" "predefined_rules_for":
> │   45:   priority                                   = each.value.priority
> │     ├────────────────
> │     │ each.value is object with 1 attribute "name"
> │ 
> │ This object does not have an attribute named "priority".
> ╵}
> ```
> 
> I guess we need update all example's default variables, and we might need an extra `precondition` to ensure that when `var.use_for_each` is `true`, then the `priority` must be set, this `precondition` could bring the users a much clearer message on what's wrong with their variables. WDYT?

I see, I agree a `precondition` is going to be needed to help the users understand that `priority` is needed in `for_each` mode. No problem.

Let me think on the approach for the tests, is not only about modifying the default variables for every example but also modifying the `modules` the examples consume to make them expect `priority` in case `for_each` is `true`.  I don't see an easy way for this as once you define `priority` even if you make it `null` this lookup:

`priority = lookup(var.predefined_rules[count.index], "priority", 4096 - length(var.predefined_rules) + count.index)
`

is going to find it and will not apply the default value, it will apply `null` and the deployment will fail. And then there is the `ActiveDirectory` module with a large list of rules...

juanjojulian commented:
Finally got some time for this... Added a precondition for priority attribute in custom rules and predefined rules when using for_each. Also modified the tests for the new code.

lonegunmanb commented on main.tf:
```diff
@@ -55,6 +55,13 @@ resource "azurerm_network_security_rule" "predefined_rules_for" {
+
+  lifecycle {
+    precondition {
+      condition     = contains(keys(each.value), "priority")
```
I've tried the following code in example:

```hcl
predefined_rules = [
    {
      name     = "Cassandra"
      priority = null
    },
  ]
```

The error message is:

```text
Error: Missing required argument
│ 
│   with module.nsg.module.nsg.azurerm_network_security_rule.predefined_rules_for["Cassandra"],
│   on ../../main.tf line 45, in resource "azurerm_network_security_rule" "predefined_rules_for":
│   45:   priority                                   = each.value.priority
│ 
│ The argument "priority" is required, but no definition was found.
```

The `priority` key exists with `null` value could bypass your `precondition`, so I would recommend:

```hcl
condition     = try(each.value.priority >= 100 && each.value.priority <= 4096, false)
```

Which would response the following error message to `null` priority:

```text
Error: Resource precondition failed
│ 
│   on ../../main.tf line 61, in resource "azurerm_network_security_rule" "predefined_rules_for":
│   61:       condition     = try(each.value.priority >= 100 && each.value.priority <= 4096, false)
│     ├────────────────
│     │ each.value.priority is null
│ 
│ Precondition failed: 'predefined_rules.priority' must be provided for predefined rules if 'var.use_for_each' is set to
│ true.
```

We can also tell our users that the priority must between [100-4096](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/network_security_rule#priority).

juanjojulian replied:
Good catch, modified the precondition with you suggestion.

lonegunmanb reviewed (changes requested):
Thanks @juanjojulian for the update, I think we can improve the precondition a little bit.

lonegunmanb commented:
Thanks for update @juanjojulian , it looks like we've met a concurrent test issue while executing version upgrade test. I'll try a fix, please wait for my test framework patch, then merge to the latest main branch, thanks for your understanding!

juanjojulian commented:
> We've got a concurrent issue when test this pr, would you please merge your branch with the latest main branch and try again? Apology for the inconvenience.

No problem. Done!

juanjojulian commented:
Hi @lonegunmanb, `version-upgrade test` is failing, if I'm not mistaken this test deploys a resource with the current module version (4.0.1) and then tries to update the module to the one in this PR. The current module version doesn't have a `use_for_each` variable declared and unfortunately Terraform don't allow us to pass command line variables if they are not declared in the code.

![CleanShot 2023-03-01 at 08 01 59@2x](https://user-images.githubusercontent.com/43988825/222067439-9592f268-cd10-4acb-88bf-de9d57d0a86f.png)

Maybe we can remove the `for_each` option from the upgrade test. As version 4.0.1 only supports `count` method there is no need to test the upgrade with `for_each`, that's not an scenario that is going to happen, and introduce it again for next version. What do you think?.

lonegunmanb commented:
> Hi @lonegunmanb, `version-upgrade test` is failing, if I'm not mistaken this test deploys a resource with the current module version (4.0.1) and then tries to update the module to the one in this PR. The current module version doesn't have a `use_for_each` variable declared and unfortunately Terraform don't allow us to pass command line variables if they are not declared in the code.
> 
> ![CleanShot 2023-03-01 at 08 01 59@2x](https://user-images.githubusercontent.com/43988825/222067439-9592f268-cd10-4acb-88bf-de9d57d0a86f.png)
> 
> Maybe we can remove the `for_each` option from the upgrade test. As version 4.0.1 only supports `count` method there is no need to test the upgrade with `for_each`, that's not an scenario that is going to happen, and introduce it again for next version. What do you think?.

Thanks @juanjojulian, would you please set `for_each` value via environment value?

```go
os.Setenv("TF_VAR_for_each", strconv.FormatBool(forEach))
```

Terraform would only try to read variable's value from env when there's no where to get it's value, so this environment won't bother previous version.

The blocker here is once we've done that, we cannot run our tests in parallel since the environment is global. Now test framework doesn't support turning on and off the parallelism.

I'd like to add a new toggle for test framework so we can turn off the parallelism, then we can use this env var approach. Apology for your bad experience, and thanks for your patience, you have improved our CI pipeline and test framework.

lonegunmanb reviewed (approved):
Thanks @juanjojulian, LGTM. 🚀

juanjojulian commented:
Thanks @lonegunmanb, no problem from my side, I'm happy to help when I have time to. In fact already applied and tested locally your suggestion for the version-upgrade test and seems to work nicely. Let me know if you want me to submit a new PR with it, find the changes bellow:

```
for _, f := range useForEach {
  os.Setenv("TF_VAR_use_for_each", strconv.FormatBool(f))
  t.Run(fmt.Sprintf("%s-%t", directory.Name(), f), func(t *testing.T) {
    test_helper.ModuleUpgradeTest(t, "Azure", "terraform-azurerm-network-security-group", fmt.Sprintf("examples/%s", d.Name()), currentRoot, terraform.Options{
      Upgrade: true,
    }, currentMajorVersion)
  })
}
```
